docs/current_docs/features/shell.mdx

```diff
@@ -0,0 +1,183 @@
+  | with-exec -- curl -L https://dagger.io \
+  | stdout
+EOF
+```
+</TabItem>
+<TabItem value="Dagger Shell">
+```shell title="Type 'dagger' to start Dagger Shell in interactive mode and enter the command below."
+container | from alpine | with-exec apk add curl | with-exec -- curl -L https://dagger.io | stdout
+```
+</TabItem>
+</Tabs>
+
+![Dagger Shell](/img/current_docs/features/shell-curl.gif)
+
+Here's another, more complex example:
+
+<Tabs groupId="shell">
+<TabItem value="System shell">
+```shell
+dagger <<EOF
+container \
+  | from cgr.dev/chainguard/wolfi-base \
+  | with-exec apk add go \
+  | with-directory /src https://github.com/golang/example#master \
+  | with-workdir /src/hello \
+  | with-exec -- go build -o hello . \
+  | file ./hello \
+  | export ./hello-from-dagger
+EOF
+```
+</TabItem>
+<TabItem value="Dagger Shell">
+```shell title="Type 'dagger' to start Dagger Shell in interactive mode and enter the command below."
+container \
+  | from cgr.dev/chainguard/wolfi-base \
+  | with-exec apk add go \
+  | with-directory /src https://github.com/golang/example#master \
+  | with-workdir /src/hello \
+  | with-exec -- go build -o hello . \
+  | file ./hello \
+  | export ./hello-from-dagger
+```
+</TabItem>
+</Tabs>
+
+![Dagger Shell](/img/current_docs/features/shell-build.gif)
+
+:::note
+Dagger Shell allows you to execute Dagger Functions as native shell commands, using a Bash-compatible syntax. However, instead of calling a subprocess for execution, Dagger Shell directly interprets commands within the Dagger CLI and executes them in containers. This means that all your commands execute as API calls in a secure, sandboxed environment, rather than as processes with direct access to the host system.
+:::
+
+## Modes of execution
+
+Dagger Shell supports multiple ways to run commands:
+
+<Tabs groupId="shell">
+<TabItem value="Inline execution">
+```shell
+dagger -c 'container | from alpine | terminal'
+```
+</TabItem>
+<TabItem value="Standard input">
+```shell title="Type 'dagger' to start Dagger Shell in interactive mode and enter the command below."
+echo 'container | from alpine | terminal' | dagger
+```
+</TabItem>
+<TabItem value="Script">
+```shell
+echo 'container | from alpine | with-exec cat /etc/os-release | stdout' > script.sh
+dagger script.sh
+```
+</TabItem>
+<TabItem value="Interactive REPL">
+```shell title="Type 'dagger' to start Dagger Shell in interactive mode and enter the command below."
+container | from alpine | terminal
+```
+</TabItem>
+</Tabs>
+
+## Built-in help
+
+Dagger Shell provides extensive built-in help:
+
+```shell
+.help             # Shows available commands
+.help MODULE      # Shows description, arguments, and return type for module constructor
+.help FUNCTION    # Shows description, arguments, and return type for function
+MODULE | .help    # Shows available functions for constructor returned by MODULE
+FUNCTION | .help  # Shows available functions for object returned by FUNCTION
```

Comment:
Note that our CLI use the `<module>` convention rather than `MODULE` when describing usage syntax.

```suggestion
.help               # Shows available commands
.help <module>      # Shows description, entrypoint arguments, and available functions
.help <function>    # Shows description, arguments, and return type for function
<function> | .help  # Shows available functions for object returned by <function>
```